### DISCUSSION TITLE
useTVEventHandler missing features and Android bug

### DISCUSSION THREAD

**Discussion by SebastienChauvin** [react-native-tvos/react-native-tvos#612] [Feature Requests]
### Description

When writing an Android TV App handling key events (for the player screen). I am missing 3 features from the `useTVEventHandler`: 

# Issues
## (1) Handler is not called when no React Native View is in focus
We have a player screen with no UI visible. In this case our View tree consists only of non focusable views. We would like however to be able to handle keys so that the UI can be displayed again

## (2) There is no "stopPropagation"
Still in this player screen we would like to do a "play/pause" using the select button while the UI is hidden. What happens is that if we do get the TVEvent evenType="select" and the play/pause is focused we have no way to prevent the `onPress` event on the focused View in the `useTVEventHandler` callback.

## (3) Some keys are filtered out
Only known KEY_EVENTS_ACTIONS are passed down to JS and unknown keys cannot be handled in JS.

# Workaround
## (1) 
Having a invisible but focusable view 

## (2)
In the `useTVEventHandler` _assume_ that the playpause button is or will be focused and do not handle play pause from there (contrary to what we do in tvOS)

## (3) 
PR and or compile from source and add keys to `KEY_EVENTS_ACTIONS`

# Suggestions
## (1)
Do key handling in the `ReactActivity` instead of ReactRootView 

## (2)
Add a stopPropagation method similar to ReactJS

## (3)
Pass all key events to JS using the integer value of the key down to JS when the key is not part of known key events. This would facilitate support of new remote controls without going through a PR in `ReactAndroidHWInputDeviceHelper` for each new key value to be supported.

### React Native Version

0.72.11

### Output of `npx react-native info`

--

### Steps to reproduce

--

### Snack, code example, screenshot, or link to a repository

--

**Comment by douglowder** [comment]
Thanks @SebastienChauvin ! 

Issue (1) was discussed in https://github.com/react-native-tvos/react-native-tvos/issues/584 and @yusufyildirim came to the same conclusion -- we need the event handling to be activity based.

I'll have a look at (2). Is anything needed for this on Apple TV?

(3) is certainly possible, provided we aren't sending too many events and overwhelming the event emitter.

Are there any other features we need related to this? I've seen a request for event throttling (for example, when rapidly scrolling through a long list of focusables, not emitting all the focus/blur events).

**Comment by evoactivity** [comment]
For other people finding this issue, I'm working around this roughly like so, in my actual component I use reanimated to show/hide the playerUI.

```tsx
import { useCallback } from 'react';
import { View, useTVEventHandler, HWEvent } from 'react-native';
import Video from 'react-native-video';
import { useInterfaceActions, useShowPlayerUI } from '../services/interface';
import PlayerLiveUI from '../components/player/live';

const PlayerScreen = () => {
  const showPlayerUI = useShowPlayerUI();
  const { togglePlayerUI } = useInterfaceActions();

  const myTVEventHandler = useCallback(
    (event: HWEvent) => {
      if (event && event.eventType && !['blur', 'focus'].includes(event.eventType)) {
        togglePlayerUI(true);
      }
    },
    [],
  );

  useTVEventHandler(myTVEventHandler);

  return (
    <>
      <Video
        source={{
          uri: player.url,
          headers: {
            'User-Agent': userAgent,
          }
        }} 
      />

      <PlayerLiveUI show={showPlayerUI} />
    </>
  );
};

export default PlayerScreen;
```

```tsx
// PlayerLiveUI
import React, { useEffect } from 'react';
import { View, Pressable, TVFocusGuideView, FocusGuideMethods } from 'react-native';

type PlayerLiveUIProps = {
  show: boolean;
};

const PlayerLiveUI: React.FC<PlayerLiveUIProps> = ({ show }) => {
  const playerUI = React.useRef<View & FocusGuideMethods>(null);
  const invisibleFocusable = React.useRef<View & FocusGuideMethods>(null);

  useEffect(() => {
    if (show) {
      playerUI.current?.requestTVFocus();
      return;
    }

    invisibleFocusable.current?.requestTVFocus();
  }, [show]);

  return (
    <>
      {/* When no focusables are on screen android doesn't fire events,
          so this TVFocusGuideView area is used so android will fire events
          it is made unfocusable when the UI is visible
          https://github.com/react-native-tvos/react-native-tvos/issues/584 */}
      <TVFocusGuideView focusable={!show} ref={invisibleFocusable}>
        <Pressable>
          <View style={{ width: 1, height: 1 }}></View>
        </Pressable>
      </TVFocusGuideView>

      <TVFocusGuideView autoFocus focusable={show} ref={playerUI}>
        <View style={{ flexDirection: 'row' }}>
          {/* player UI here with other pressables */}
        </View>
      </TVFocusGuideView>
    </>
  );
};

export default PlayerLiveUI;
```

**Comment by hamdalf** [comment]
Hello,

I would like to ask about 'stopPropagation'. Because it's really needed for the apps on 'AndroidTV embedded on real TV'. (ex. Sony Bravia)

One example is 'Channel_Up/Down' button case. When a user press 'Channel_up' button on the remote controller, the app can use the event. However, within a second, TV system's 'channel setup' screen shows. If there is an app which registered as 'live TV app', the previous app (which received that Channel_up event first) is closed and the live TV app is launched.

So, the app needs a way to prevent event propagation to system for preventing app closing.

With native app development (without React Native), we can easily prevent event propagation. However, in react-native-tvos, when we are using 'useTVEventHandler()', there is no way to stop event propagation.

Number keys are also same. It will open live TV of the system also.

So, if there is a way to prevent event propagation, it will be very useful for the apps on real TV which has more keys on remote controller.
